netebpfext/user/fwp_um.cpp

```diff
@@ -68,6 +70,203 @@ _fwp_engine::classify_test_packet(_In_ const GUID* layer_guid, NET_IFINDEX if_in
     return result.actionType;
 }
 
+const uint32_t _test_destination_ipv4_address = 0x01020304;
+static FWP_BYTE_ARRAY16 _test_destination_ipv6_address = {1, 2, 3, 4};
+const uint16_t _test_destination_port = 1234;
+const uint32_t _test_source_ipv4_address = 0x05060708;
+static FWP_BYTE_ARRAY16 _test_source_ipv6_address = {5, 6, 7, 8};
+const uint16_t _test_source_port = 5678;
+const uint8_t _test_protocol = IPPROTO_TCP;
+const uint32_t _test_compartment_id = 1;
+static FWP_BYTE_BLOB _test_app_id = {.size = 2, .data = (uint8_t*)"\\"};
+static uint64_t _test_interface_luid = 1;
+
+// This is used to test the bind hook.
+FWP_ACTION_TYPE
+_fwp_engine::test_bind_ipv4()
+{
+    std::unique_lock l(lock);
+
+    FWPS_INCOMING_VALUE0 incoming_value[FWPS_FIELD_ALE_RESOURCE_ASSIGNMENT_V4_MAX] = {};
+    incoming_value[FWPS_FIELD_ALE_RESOURCE_ASSIGNMENT_V4_IP_LOCAL_PORT].value.uint16 = _test_destination_port;
+    incoming_value[FWPS_FIELD_ALE_RESOURCE_ASSIGNMENT_V4_IP_LOCAL_ADDRESS].value.uint32 =
+        _test_destination_ipv4_address;
+    incoming_value[FWPS_FIELD_ALE_RESOURCE_ASSIGNMENT_V4_IP_PROTOCOL].value.uint8 = _test_protocol;
+    incoming_value[FWPS_FIELD_ALE_RESOURCE_ASSIGNMENT_V4_ALE_APP_ID].value.byteBlob = &_test_app_id;
+
+    return test_callout(FWPS_LAYER_ALE_RESOURCE_ASSIGNMENT_V4, FWPM_LAYER_ALE_RESOURCE_ASSIGNMENT_V4, incoming_value);
```

Comment:
we're calling ```test_callout``` with the the ```_fwp_engine::lock``` held (above).  This function seems to do a _lot_ of possibly un-bounded work (call other functions) and I was wondering if we needed the lock for all that time.  Was also curious about the potential for an (indirect) deadlock in some subsequent code-path(s).

Reply:
that was already the precedent in this file (see line 15) so is not changed by this PR.  If it is an issue, it was already an issue and should be tracked by a different bug.

Reply:
Filed an [issue](https://github.com/microsoft/ebpf-for-windows/issues/1898) to address this.